server-mirror-git/src/main/java/com/linecorp/centraldogma/server/internal/mirror/AbstractGitMirror.java

```diff
@@ -131,283 +136,285 @@ public GitMirror(Cron schedule, MirrorDirection direction, MirrorCredential cred
+            }
+        });
 
-            remoteSetUrl.setUriType(UriType.PUSH);
-            remoteSetUrl.call();
+        executor.execute(Command.push(
+                MIRROR_AUTHOR, localRepo().parent().name(), localRepo().name(),
+                Revision.HEAD, summary, detail, Markup.PLAINTEXT, changes.values())).join();
+    }
 
-            // XXX(trustin): Do not garbage-collect a Git repository while the server is serving the clients
-            //               because GC can incur large amount of disk writes that slow the server.
-            //               Ideally, we could introduce some sort of maintenance mode.
-            // Keep things clean.
-            //git.gc().call();
+    private boolean needsFetch(Ref headBranchRef, String mirrorStatePath, Revision localRev)
+            throws JsonParseException, JsonMappingException {
+        final Entry<?> mirrorState = localRepo().getOrNull(localRev, mirrorStatePath).join();
+        final String localSourceRevision;
+        if (mirrorState == null || mirrorState.type() != EntryType.JSON) {
+            localSourceRevision = null;
+        } else {
+            localSourceRevision = Jackson.treeToValue((TreeNode) mirrorState.content(),
+                                                      MirrorState.class).sourceRevision();
+        }
 
-            success = true;
-            return git;
-        } finally {
-            if (!success) {
-                git.close();
-            }
+        final ObjectId headCommitId = headBranchRef.getObjectId();
+        if (headCommitId.name().equals(localSourceRevision)) {
+            final String abbrId = headCommitId.abbreviate(OBJECT_ID_ABBREV_STRING_LENGTH).name();
+            logger.info("Repository '{}' already at {}, {}#{}", localRepo().name(), abbrId,
+                        remoteRepoUri(), remoteBranch());
+            return false;
         }
+        return true;
     }
 
-    private String getHeadBranchRefName(GitWithAuth git) throws GitAPIException {
-        // Use the given branch if available.
+    private Ref getHeadBranchRef(
+            GitWithAuth git, Consumer<TransportCommand<?,?>> configurator) throws GitAPIException {
         if (remoteBranch() != null) {
-            return Constants.R_HEADS + remoteBranch();
+            final String headBranchRefName = Constants.R_HEADS + remoteBranch();
+            final Collection<Ref> refs = lsRemote(git, configurator, true);
+            return findHeadBranchRef(git, headBranchRefName, refs);
         }
 
         // Otherwise, we need to figure out which branch we should fetch.
         // Fetch the remote reference list to determine the default branch.
-        final Collection<Ref> refs = git.lsRemote()
-                                        .setTags(false)
-                                        .setTimeout(GIT_TIMEOUT_SECS)
-                                        .call();
+        final Collection<Ref> refs = lsRemote(git, configurator, false);
```

Comment:
Question) I know this is just a refactor, but I'm curious if automatic branch detection actually works with the new logic. If heads is false, IIUC `refs/heads/*` aren't fetched which means branches aren't fetched.

I have no problem if this works, but wanted to point out just in case it was somehow missed

Reply:
> If heads is false, IIUC refs/heads/* aren't fetched which means branches aren't fetched.

It's fetched. 😉 
```
// head true:
centraldogma-line git:(master) git ls-remote --heads
From https://git.linecorp.com/minu-song/centraldogma-line.git
6fa34705f21f9b032e71b0a14069ec1d456ea84f	refs/heads/add_actions
0be19a8460f987a1eea8dd621c32d45c1a144678	refs/heads/master

// head false:
centraldogma-line git:(master) git ls-remote
From https://git.linecorp.com/minu-song/centraldogma-line.git
0be19a8460f987a1eea8dd621c32d45c1a144678	HEAD
6fa34705f21f9b032e71b0a14069ec1d456ea84f	refs/heads/add_actions
0be19a8460f987a1eea8dd621c32d45c1a144678	refs/heads/master
1cd207f84ce0ce83e459bf2d355593f0ac44e17b	refs/heads/use_Javet
e73e4e573cd6183a00bb1fdf8f85d7b69f821dc7	refs/tags/centraldogma-line-0.14.0.0
3be144a526e31513cbc720b8cc8a63a2db7867c4	refs/tags/centraldogma-line-0.14.1.0
...
```